presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -0,0 +1,374 @@
+                return expression;
+            }
+
+            Object literalValue = ExpressionInterpreter.expressionOptimizer(expression.getRight(), metadata, session, typeAnalyzer.getTypes(session, types, expression.getRight()))
+                    .optimize(NoOpSymbolResolver.INSTANCE);
+
+            if (literalValue instanceof Expression) {
+                return expression;
+            }
+
+            Cast cast = (Cast) expression.getLeft();
+            ComparisonExpression.Operator operator = expression.getOperator();
+
+            if (literalValue == null) {
+                switch (operator) {
+                    case EQUAL:
+                    case NOT_EQUAL:
+                    case LESS_THAN:
+                    case LESS_THAN_OR_EQUAL:
+                    case GREATER_THAN:
+                    case GREATER_THAN_OR_EQUAL:
+                        return new Cast(new NullLiteral(), BooleanType.BOOLEAN.toString());
+                    case IS_DISTINCT_FROM:
+                        return TRUE_LITERAL;
+                    default:
+                        throw new UnsupportedOperationException("not yet implemented");
+                }
+            }
+
+            Type sourceType = typeAnalyzer.getType(session, types, cast.getExpression());
+            Type literalType = typeAnalyzer.getType(session, types, expression.getRight());
+
+            if (sourceType.equals(BIGINT) && literalType.equals(DOUBLE) ||
+                    sourceType.equals(BIGINT) && literalType.equals(REAL) ||
+                    sourceType.equals(INTEGER) && literalType.equals(REAL)) {
+                // Special case: implicit casts between BIGINT->DOUBLE/REAL and INTEGER->REAL are broken
+                // as they are not injective. Skip to avoid producing unexpected results
+                return expression;
+            }
+
+            if (!metadata.getTypeManager().canCoerce(sourceType, literalType)) {
+                // transformation is only valid when it's an implicit CAST
+                return expression;
+            }
+
+            Signature sourceToTarget = metadata.getFunctionRegistry().getCoercion(sourceType, literalType);
+
+            Optional<Type.Range> range = sourceType.getRange();
+            if (range.isPresent()) {
+                Object max = range.get().getMax();
+                Object maxInTargetType = functionInvoker.invoke(
+                        sourceToTarget,
+                        session.toConnectorSession(),
+                        max);
+
+                int upperBoundComparison = compare(literalType, literalValue, maxInTargetType);
+                if (upperBoundComparison > 0) {
+                    // larger than maximum representable value
+                    switch (operator) {
+                        case EQUAL:
+                        case GREATER_THAN:
+                        case GREATER_THAN_OR_EQUAL:
+                            return falseIfNotNull(cast.getExpression());
```

Comment:
Let source type be `REAL`, target type be `DOUBLE` and `literal` be `Infinity` (`DOUBLE`).
For `cast.getExpression()` being infinity (`REAL`), the new predicate is not equivalent.

Reply:
Hah, you read my mind. I came to the same conclusion and have been working on a fix. Long story short, real and double don’t have a well-defined range.